lunabot_control/src/mpc.cpp

```diff
@@ -229,33 +191,43 @@ double MPC::calculate_cost_(Eigen::MatrixXd rollout) {
   double cost = 0;
   std::vector<double> robot_pos = this->robot_pos_;
   std::vector<double> goal = this->goal_;
+#pragma omp parallel for reduction(+ : cost)
   for (int i = 0; i < rollout.rows(); ++i) {
     Eigen::RowVectorXd position = rollout.row(i);
     cost += this->w_linear_ * ((position(0) - robot_pos[0]) * (position(0) - robot_pos[0]) +
                                (position(1) - robot_pos[1]) * (position(1) - robot_pos[1]));
     cost += this->w_angular_ * (position(2) - robot_pos[2]) * (position(2) - robot_pos[2]);
     int path_cost_i = std::min(path_ind_ + i, (int)path_.size() - 1);
-    cost += this->w_waypoint_ *
+    cost += this->w_waypoint_pos_ *
             ((position(0) - path_[path_cost_i][0]) * (position(0) - path_[path_cost_i][0]) +
              (position(1) - path_[path_cost_i][1]) * (position(1) - path_[path_cost_i][1]));
+
+    cost += this->w_waypoint_rot_ * (position(2) - path_[path_cost_i][2]) *
```

Comment:
We may want to redo how we calculate cost / reward for path following if we used a smooth path. Idk how many more points you generate but we may be constraining ourselves too strictly to the path with the current method. Not sure how results look tho.